codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java

```diff
@@ -417,17 +418,229 @@ public void windowUpdateFrameDecrementsConsumedBytes() throws Exception {
+                        listenerExecuted.setSuccess(null);
+                    }
+                }
+        );
+        ByteBuf data = Unpooled.buffer().writeZero(100);
+        ChannelFuture f = channel.writeAndFlush(new DefaultHttp2DataFrame(data).stream(stream));
+        assertTrue(f.isSuccess());
+
+        listenerExecuted.syncUninterruptibly();
+        assertTrue(listenerExecuted.isSuccess());
+    }
+
+    @Test
+    public void newOutboundStreamsShouldBeBuffered() throws Exception {
+        setUp(Http2FrameCodecBuilder.forServer().encoderEnforceMaxConcurrentStreams(true),
+              new Http2Settings().maxConcurrentStreams(1));
+
+        Http2FrameStream stream1 = frameCodec.newStream();
+        Http2FrameStream stream2 = frameCodec.newStream();
+
+        ChannelPromise promise1 = channel.newPromise();
+        ChannelPromise promise2 = channel.newPromise();
+
+        channel.writeAndFlush(new DefaultHttp2HeadersFrame(new DefaultHttp2Headers()).stream(stream1), promise1);
+        channel.writeAndFlush(new DefaultHttp2HeadersFrame(new DefaultHttp2Headers()).stream(stream2), promise2);
+
+        assertTrue(isStreamIdValid(stream1.id()));
+        channel.runPendingTasks();
+        assertTrue(isStreamIdValid(stream2.id()));
+
+        assertTrue(promise1.syncUninterruptibly().isSuccess());
+        assertFalse(promise2.isDone());
+
+        // Increase concurrent streams limit to 2
+        frameListener.onSettingsRead(http2HandlerCtx, new Http2Settings().maxConcurrentStreams(2));
+
+        channel.flush();
+
+        assertTrue(promise2.syncUninterruptibly().isSuccess());
+    }
+
+    @Test
+    public void streamIdentifiersExhausted() throws Http2Exception {
+        int maxServerStreamId = Integer.MAX_VALUE - 1;
+
+        assertNotNull(frameCodec.connection().local().createStream(maxServerStreamId, false));
+
+        Http2FrameStream stream = frameCodec.newStream();
+        assertNotNull(stream);
+
+        ChannelPromise writePromise = channel.newPromise();
+        channel.writeAndFlush(new DefaultHttp2HeadersFrame(new DefaultHttp2Headers()).stream(stream), writePromise);
+
+        assertThat(writePromise.cause(), instanceOf(Http2NoMoreStreamIdsException.class));
+    }
+
+    @Test
+    public void receivePing() throws Http2Exception {
+        ByteBuf data = Unpooled.buffer(8).writeLong(12345);
+        frameListener.onPingRead(http2HandlerCtx, releaseLater(data));
+
+        Http2PingFrame pingFrame = inboundHandler.readInbound();
+        assertNotNull(pingFrame);
+
+        assertEquals(data, pingFrame.content());
+        assertFalse(pingFrame.ack());
+        pingFrame.release();
+    }
+
+    @Test
+    public void sendPing() {
+        ByteBuf data = Unpooled.buffer(8).writeLong(12345);
+        channel.writeAndFlush(new DefaultHttp2PingFrame(data));
+
+        verify(frameWriter).writePing(eq(http2HandlerCtx), eq(false), eq(data), anyChannelPromise());
```

Comment:
ahh yes this is an embeddedchannel and you call `finishAndReleaseAll` so all good.